scalable-go-service/index.md

```diff
@@ -0,0 +1,322 @@
+	Count         int
+	In            chan RateRQ
+	TokenRQChan   chan<- TokenRQ
+	stop          chan struct{}
+}
+
+// Start initializes and starts n unique rate loader based on the Count parameter.
+// Each rate loader listens for requests on the pool's In channel.
+// For each request, a loader asks for a token through the TokenRQChan,
+// gets rates from the vendor, and puts the result on the requests'
+// desired response channel (RateRSChan).
+func (rlp *RateLoaderPool) Start() {
+	rlp.stop = make(chan struct{}, 1)
+	var wg sync.WaitGroup
+	for i := 0; i < rlp.Count; i++ {
+		wg.Add(1)
+		tokenRSChan := make(chan TokenRS)
+		go func() {
+		loop:
+			for {
+				select {
+				case rq := <-rlp.In:
+
+					// ask for a token from a token loader
+					rlp.TokenRQChan <- TokenRQ{...}
+					token := <-tokenRSChan
+
+					// get rates for a hotel
+					r := getRateFromVendor(...)
+
+					rq.RateRSChan <- RateRS{r}
+				case <-rlp.stop:
+					break loop
+				}
+			}
+			wg.Done()
+		}()
+	}
+    // blocks until all workers in the pool have stopped
+	wg.Wait()
+}
+
+// Stop stops all workers in the pool
+func (rlp *RateLoaderPool) Stop() {
+	close(rlp.stop)
+}
+```
+
+#### Token Loader
+
+```go
+type TokenLoader struct {
+	In   <-chan TokenRQ
+	stop chan struct{}
+}
+
+// Start initializes and starts the token loader
+func (tl *TokenLoader) Start() {
+	tl.stop = make(chan struct{}, 1)
+	var token *TokenRS
+loop:
+	for {
+		select {
+		case rq := <-tl.In:
+			rq.TokenRSChan <- getAuthenticationToken(...)
+		case <-tl.stop:
+			break loop
+		}
+	}
+}
+
+// Stop token loader
+func (tl *TokenLoader) Stop() {
+	close(tl.stop)
+}
+```
+
+## Importance of Communication
+
+As these vendor services grow, the team has had discussions around how much we
+should stick to the worker model as we add features. The debate usually revolves
+around whether another worker or pool of workers is required to add small
+features, which would normally be implemented as helper functions. For example,
+a recent feature we added was the ability to log all interactions with the
+vendors' APIs to S3. This feature was initially implemented with a simple helper
+function that would take a JSON object and log it in our S3 bucket. To ensure
+that the request to S3 did not block the regular flow, we called the helper
+function in a `goroutine`.
+
+This solution was not consistent with the rest of the system. Conceptually we
+had added a component that could execute independently but effectively
+communicating that as we did with other concurrent modules. The failure to
+communicate clearly made this part of the program hard to follow and it would be
+confusing to new engineers. Furthermore, we had moved away from the paradigm
+identified as optimum for the problem. It would be difficult to accurately
+assess the overall solution if we do not consistently follow and communicate the
+strategy.
+
+## Helpful Links
+
+<https://blog.golang.org/share-memory-by-communicating><br/>
```

Comment:
This should actually be a linked list, not just different lines.

Reply:
Even better, inline these references as links in the main body of the post rather than an appendix at the end.